### PR TITLE
Sensor EEH210 support

### PR DESCRIPTION
**Pull request by resnikal71202**
# What does this implement/fix?

add support for the [EEH210 sensor](https://www.epluse.com/fileadmin/data/product/eehx10/datasheet_EEH210.pdf)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** [https://github.com/esphome/esphome-docs/pull/3036](https://github.com/esphome/esphome-docs/pull/3036)

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
external_components:
  - source:
      type: git
      url: https://github.com/resnikal71202/esphome.git
      ref: dev
    components: [ eeh210 ]

sensor:
- platform: eeh210
  address: 0x40
  temperature:
    name: "Temperature"
  humidity:
    name: "Humidity"
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/eeh210/eeh210.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+static const uint8_t EEH210_I2C_COMMAND_TRIGGER_T_RH_MEASURE_NO_HOLD = 0xF1;
+static const uint8_t EEH210_I2C_COMMAND_TRIGGER_T_MEASURE_NO_HOLD = 0xF3;
+static const uint8_t EEH210_I2C_COMMAND_TRIGGER_RH_MEASURE_NO_HOLD = 0xF5;
+
+static const uint8_t EEH210_I2C_COMMAND_WRITE_USER_REGISTER = 0xE6;
+static const uint8_t EEH210_I2C_COMMAND_READ_USER_REGISTER = 0xFE;
+
+void EEH210Component::setup() { ESP_LOGCONFIG(TAG, "Setting up EEH210..."); }
+
```
Comment (jesserockz) [inline esphome/components/eeh210/eeh210.cpp]: Pointless empty function
```suggestion
```

File: esphome/components/eeh210/eeh210.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+float EEH210Component::get_setup_priority() const { return setup_priority::DATA; }
+void EEH210Component::update() {
+  uint8_t command[] = {EEH210_I2C_COMMAND_TRIGGER_T_RH_MEASURE_HOLD};
+  this->write(command, 1, true);
+  this->set_timeout(50, [this]() {
+    uint8_t i2c_response[6];
+    this->read(i2c_response, 6);
+
+    unsigned int raw_temperature = (i2c_response[1] << 8) | (i2c_response[2] & 0b11111100);
```
Comment (jesserockz) [inline esphome/components/eeh210/eeh210.cpp]:
```suggestion
    uint32_t raw_temperature = (i2c_response[1] << 8) | (i2c_response[2] & 0b11111100);
```

File: esphome/components/eeh210/eeh210.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    float temperature = -46.85 + 175.72 * (raw_temperature / 65536.0);
+
+    unsigned int raw_humidity = (i2c_response[3] << 8) | (i2c_response[4] & 0b11111100);
```
Comment (jesserockz) [inline esphome/components/eeh210/eeh210.cpp]:
```suggestion
    uint32_t raw_humidity = (i2c_response[3] << 8) | (i2c_response[4] & 0b11111100);
```

File: esphome/components/eeh210/eeh210.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    if (this->humidity_sensor_ != nullptr)
+      this->humidity_sensor_->publish_state(relative_humidity);
+    this->status_clear_warning();
+  });
+}
+
+unsigned char EEH210Component::calc_crc8_(const unsigned char buf[], unsigned char from, unsigned char to) {
+  unsigned char crc_val = 0xFF;
+  unsigned char i = 0;
+  unsigned char j = 0;
```
Comment (jesserockz) [inline esphome/components/eeh210/eeh210.cpp]: We prefer to use `uint8_t` in ESPHome.
```suggestion
uint8_t EEH210Component::calc_crc8_(const uint8_t buf[], uint8_t from, uint8_t to) {
  uint8_t crc_val = 0xFF;
  uint8_t i = 0;
  uint8_t j = 0;
```

File: esphome/components/eeh210/eeh210.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { temperature_sensor_ = temperature_sensor; }
+  void set_humidity_sensor(sensor::Sensor *humidity_sensor) { humidity_sensor_ = humidity_sensor; }
+
+  float get_setup_priority() const override;
+  void setup() override;
+  void dump_config() override;
+  void update() override;
+
+ protected:
+  unsigned char calc_crc8_(const unsigned char buf[], unsigned char from, unsigned char to);
```
Comment (jesserockz) [inline esphome/components/eeh210/eeh210.h]:
```suggestion
  uint8_t calc_crc8_(const uint8_t buf[], uint8_t from, uint8_t to);
```

File: esphome/components/eeh210/eeh210.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+namespace eeh210 {
+
+/// This class implements support for the hte501 of temperature i2c sensors.
+class EEH210Component : public PollingComponent, public i2c::I2CDevice {
+ public:
+  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { temperature_sensor_ = temperature_sensor; }
+  void set_humidity_sensor(sensor::Sensor *humidity_sensor) { humidity_sensor_ = humidity_sensor; }
+
+  float get_setup_priority() const override;
+  void setup() override;
```
Comment (jesserockz) [inline esphome/components/eeh210/eeh210.h]: Not needed
```suggestion
```

File: esphome/components/eeh210/sensor.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+
+    if CONF_TEMPERATURE in config:
+        conf = config[CONF_TEMPERATURE]
+        sens = await sensor.new_sensor(conf)
```
Comment (jesserockz) [inline esphome/components/eeh210/sensor.py]:
```suggestion
    if temperature_config := config.get(CONF_TEMPERATURE):
        sens = await sensor.new_sensor(temperature_config)
```

File: esphome/components/eeh210/sensor.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        cg.add(var.set_temperature_sensor(sens))
+
+    if CONF_HUMIDITY in config:
+        conf = config[CONF_HUMIDITY]
+        sens = await sensor.new_sensor(conf)
```
Comment (jesserockz) [inline esphome/components/eeh210/sensor.py]:
```suggestion
    if humidity_config := config.get(CONF_HUMIDITY):
        sens = await sensor.new_sensor(humidity_config)
```

File: esphome/components/eeh210/eeh210.cpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+float EEH210Component::get_setup_priority() const { return setup_priority::DATA; }
+void EEH210Component::update() {
+  uint8_t command[] = {EEH210_I2C_COMMAND_TRIGGER_T_RH_MEASURE_HOLD};
+  this->write(command, 1, true);
+  this->set_timeout(50, [this]() {
+    uint8_t i2c_response[6];
+    this->read(i2c_response, 6);
+
+    unsigned int raw_temperature = (i2c_response[1] << 8) | (i2c_response[2] & 0b11111100);
+    float temperature = -46.85 + 175.72 * (raw_temperature / 65536.0);
```
Comment (jesserockz) [inline esphome/components/eeh210/eeh210.cpp]: We want floats, not doubles =)
```suggestion
    float temperature = -46.85f + 175.72f * (raw_temperature / 65536.0f);
```

File: esphome/components/eeh210/eeh210.cpp [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+    unsigned int raw_humidity = (i2c_response[3] << 8) | (i2c_response[4] & 0b11111100);
+    float relative_humidity = -6.0 + 125.0 * (raw_humidity / 65536.0);
```
Comment (jesserockz) [inline esphome/components/eeh210/eeh210.cpp]: Same here
```suggestion
    float relative_humidity = -6.0f + 125.0f * (raw_humidity / 65536.0f);
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by resnikal71202 — COMMENTED** [review]
tested on an esp32:

```
esphome:
  name: esphome-web-eeh210
  friendly_name: EEH210 Satelite

esp32:
  board: esp32dev
  framework:
    type: arduino

i2c:
  id: i2c_0
  sda: GPIO21
  scl: GPIO22
  frequency: 50kHz
  scan: False

logger:

api:
  encryption:
    key: "..."

ota:


wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

captive_portal:
    
external_components:
  - source:
      type: git
      url: https://github.com/resnikal71202/esphome.git
      ref: dev
    components: [ eeh210 ]

sensor:
- platform: eeh210
  address: 0x40
  temperature:
    name: "Temperature"
  humidity:
    name: "Humidity"

# deep_sleep:
#   run_duration: 15s
#   sleep_duration: 1min
```
log:
```
INFO ESPHome 2023.11.4
INFO Reading configuration /config/esphome/esphome-web-114a70.yaml...
INFO Starting log output from esphome-web-eeh210.local using esphome API
INFO Successfully connected to esphome-web-eeh210 in 0.493s
INFO Successful handshake with esphome-web-eeh210 in 0.116s
[02:12:37][I][app:102]: ESPHome version 2023.11.4 compiled on Nov 25 2023, 02:05:18
[02:12:37][C][wifi:559]: WiFi:
[02:12:37][C][wifi:391]:   Local MAC: 78:E3:6D:0A:66:58
[02:12:37][C][wifi:396]:   SSID: '...'[redacted]
[02:12:37][C][wifi:397]:   IP Address: 192.168.0.98
[02:12:37][C][wifi:399]:   BSSID: DC:15:C8:E2:08:3D[redacted]
[02:12:37][C][wifi:400]:   Hostname: 'esphome-web-eeh210'
[02:12:37][C][wifi:402]:   Signal strength: -71 dB ▂▄▆█
[02:12:37][C][wifi:406]:   Channel: 11
[02:12:37][C][wifi:407]:   Subnet: 255.255.255.0
[02:12:37][C][wifi:408]:   Gateway: 192.168.0.1
[02:12:37][C][wifi:409]:   DNS1: 192.168.0.1
[02:12:37][C][wifi:410]:   DNS2: 0.0.0.0
[02:12:37][C][logger:416]: Logger:
[02:12:37][C][logger:417]:   Level: DEBUG
[02:12:37][C][logger:418]:   Log Baud Rate: 115200
[02:12:37][C][logger:420]:   Hardware UART: UART0
[02:12:37][C][i2c.arduino:053]: I2C Bus:
[02:12:37][C][i2c.arduino:054]:   SDA Pin: GPIO21
[02:12:37][C][i2c.arduino:055]:   SCL Pin: GPIO22
[02:12:37][C][i2c.arduino:056]:   Frequency: 50000 Hz
[02:12:37][C][i2c.arduino:059]:   Recovery: bus successfully recovered
[02:12:38][C][eeh210.sensor:021]: EEH210:
[02:12:38][C][eeh210.sensor:022]:   Address: 0x40
[02:12:38][C][eeh210.sensor:034]:   Update Interval: 60.0s
[02:12:38][C][eeh210.sensor:035]:   Temperature 'Temperature'
[02:12:38][C][eeh210.sensor:035]:     Device Class: 'temperature'
[02:12:38][C][eeh210.sensor:035]:     State Class: 'measurement'
[02:12:38][C][eeh210.sensor:035]:     Unit of Measurement: '°C'
[02:12:38][C][eeh210.sensor:035]:     Accuracy Decimals: 4
[02:12:38][C][eeh210.sensor:036]:   Humidity 'Humidity'
[02:12:38][C][eeh210.sensor:036]:     Device Class: 'humidity'
[02:12:38][C][eeh210.sensor:036]:     State Class: 'measurement'
[02:12:38][C][eeh210.sensor:036]:     Unit of Measurement: '%'
[02:12:38][C][eeh210.sensor:036]:     Accuracy Decimals: 1
[02:12:38][C][captive_portal:088]: Captive Portal:
[02:12:38][C][mdns:115]: mDNS:
[02:12:38][C][mdns:116]:   Hostname: esphome-web-eeh210
[02:12:38][C][ota:097]: Over-The-Air Updates:
[02:12:38][C][ota:098]:   Address: esphome-web-eeh210.local:3232
[02:12:38][C][api:139]: API Server:
[02:12:38][C][api:140]:   Address: esphome-web-eeh210.local:6053
[02:12:38][C][api:142]:   Using noise encryption: YES
[02:13:36][D][eeh210.sensor:053]: Got temperature=21.6190°C humidity=56.3779%
[02:13:36][D][sensor:094]: 'Temperature': Sending state 21.61903 °C with 4 decimals of accuracy
[02:13:36][D][sensor:094]: 'Humidity': Sending state 56.37793 % with 1 decimals of accuracy
```

**Pull request closed without merging**